pkg/cmd/server/options.go

```diff
@@ -14,51 +14,75 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+// Package server provides configuration options for the metrics API server.
 package server
 
 import (
 	"fmt"
 	"net"
 
+	"github.com/spf13/pflag"
+
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
 	genericapiserver "k8s.io/apiserver/pkg/server"
 	genericoptions "k8s.io/apiserver/pkg/server/options"
 	openapicommon "k8s.io/kube-openapi/pkg/common"
 
 	"sigs.k8s.io/custom-metrics-apiserver/pkg/apiserver"
 )
 
+// CustomMetricsAdapterServerOptions contains the of options used to configure
+// the metrics API server.
+//
+// It is based on a subset of [genericoptions.RecommendedOptions].
 type CustomMetricsAdapterServerOptions struct {
-	// genericoptions.ReccomendedOptions - EtcdOptions
 	SecureServing  *genericoptions.SecureServingOptionsWithLoopback
 	Authentication *genericoptions.DelegatingAuthenticationOptions
 	Authorization  *genericoptions.DelegatingAuthorizationOptions
 	Audit          *genericoptions.AuditOptions
 	Features       *genericoptions.FeatureOptions
 
-	// OpenAPIConfig
 	OpenAPIConfig *openapicommon.Config
+	EnableMetrics bool
 }
 
+// NewCustomMetricsAdapterServerOptions creates a new instance of
+// CustomMetricsAdapterServerOptions with its default values.
 func NewCustomMetricsAdapterServerOptions() *CustomMetricsAdapterServerOptions {
 	o := &CustomMetricsAdapterServerOptions{
 		SecureServing:  genericoptions.NewSecureServingOptions().WithLoopback(),
 		Authentication: genericoptions.NewDelegatingAuthenticationOptions(),
 		Authorization:  genericoptions.NewDelegatingAuthorizationOptions(),
 		Audit:          genericoptions.NewAuditOptions(),
 		Features:       genericoptions.NewFeatureOptions(),
+
+		EnableMetrics: true,
 	}
 
 	return o
 }
 
-func (o CustomMetricsAdapterServerOptions) Validate(args []string) error {
-	return nil
+// Validate validates CustomMetricsAdapterServerOptions
+func (o CustomMetricsAdapterServerOptions) Validate() error {
```

Comment:
Actually, I used the sample-apiserver as an example (since this file seems to be "forked" from sample-apiserver), and they use this signature:

https://github.com/kubernetes/kubernetes/blob/86f3fa94f1b75da275a8040e4155840a7b473dc1/staging/src/k8s.io/sample-apiserver/pkg/cmd/server/start.go#L102

To which `Validate() []error` function do you refer?

Reply:
I was looking at the recommended options code: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/options/recommended.go#L158-L172.

Now that you've mention it is coming from the sample-apiserver I am fine with both but would prefer the recommended options approach since custom-metrics-apiserver is a library (similarly to recommended options) and not a complete apiserver.

Reply:
Ok, it makes sense. Actually, this code came from the sample-apiserver, but it was before the introduction of the builder in 35cf4d7459d1, which changed the way it is used.
I'll update the PR with `Validate() []error` and `ApplyTo`.
No problem changing the API between versions?

Reply:
> No problem changing the API between versions?

It needs to be taken into consideration since it will require users to change their implementation a bit, but still, the behavior will not change a lot, they will just need to update the implementation slightly. As long as it is described in the release note, I am fine with it.

Moreover I am thinking that we should try to bring the resource metrics API over before doing a next release so users will need to do a bit more changes this release compare to others so now should be the best time to unify the implementation IMO

Reply:
> Ok, it makes sense. Actually, this code came from the sample-apiserver, but it was before the introduction of the builder in https://github.com/kubernetes-sigs/custom-metrics-apiserver/commit/35cf4d7459d1c7894687ce0be723b5a2c65250f2, which changed the way it is used.

Make sense to use the newer implementation then.